File: frontend/src/lib/vocabulary-guard.ts
export function applyVocabularyGuard(text: string): string {
  if (!text) return text;
  
  // Case-insensitive replacements for banned terms
  let sanitized = text;
  
  const replacements = [
    { pattern: /\b(buy|buying)\b/gi, replacement: "consider" },
    { pattern: /\b(sell|selling)\b/gi, replacement: "re-evaluate" },
    { pattern: /\b(recommend|recommends|recommending|recommendation)\b/gi, replacement: "highlight" },
    { pattern: /\b(should buy|should sell)\b/gi, replacement: "might fit your criteria" },
    { pattern: /\b(best stock|top stock)\b/gi, replacement: "strong match" },
    { pattern: /\b(guaranteed|sure thing)\b/gi, replacement: "historically aligned" }
  ];
  
  for (const { pattern, replacement } of replacements) {
    sanitized = sanitized.replace(pattern, replacement);
  }
  
  return sanitized;
}

File: backend/app/utils/vocabulary_guard.py
import re

def apply_vocabulary_guard(text: str) -> str:
    if not text:
        return text
        
    replacements = [
        (r'\b(buy|buying)\b', 'consider'),
        (r'\b(sell|selling)\b', 're-evaluate'),
        (r'\b(recommend|recommends|recommending|recommendation)\b', 'highlight'),
        (r'\b(should buy|should sell)\b', 'might fit your criteria'),
        (r'\b(best stock|top stock)\b', 'strong match'),
        (r'\b(guaranteed|sure thing)\b', 'historically aligned')
    ]
    
    sanitized = text
    for pattern, repl in replacements:
        sanitized = re.sub(pattern, repl, sanitized, flags=re.IGNORECASE)
        
    return sanitized
